// ionic-angular/es2015/components/tap-click/activator.js
import { nativeTimeout, rafFrames } from '../../util/dom';
import { isActivatedDisabled } from './activator-base';
export class Activator {
    constructor(app, config) {
        this.app = app;
        this._queue = [];
        this._active = [];
        this.activatedDelay = ADD_ACTIVATED_DEFERS;
        this.clearDelay = CLEAR_STATE_DEFERS;
        this._css = config.get('activatedClass') || 'activated';
    }
    clickAction(ev, activatableEle, startCoord) {
        this._scheduleClear();
        this._queue.length = 0;
        for (var i = 0; i < this._active.length; i++) {
            this._active[i].classList.remove(this._css);
        }
        this._active.length = 0;
        if (activatableEle && activatableEle.parentNode) {
            this._active.push(activatableEle);
            activatableEle.classList.add(this._css);
        }
    }
    downAction(ev, activatableEle, startCoord) {
        if (isActivatedDisabled(ev, activatableEle)) {
            return;
        }
        this.unscheduleClear();
        this.deactivate();
        this._queue.push(activatableEle);
        this._activeRafDefer = rafFrames(this.activatedDelay, () => {
            let activatableEle;
            for (let i = 0; i < this._queue.length; i++) {
                activatableEle = this._queue[i];
                this._active.push(activatableEle);
                activatableEle.classList.add(this._css);
            }
            this._queue.length = 0;
            this._clearDeferred();
        });
    }
    upAction(ev, activatableEle, startCoord) {
        this._scheduleClear();
    }
    _scheduleClear() {
        if (this._clearRafDefer) {
            return;
        }
        this._clearRafDefer = rafFrames(this.clearDelay, () => {
            this.clearState();
            this._clearRafDefer = null;
        });
    }
    unscheduleClear() {
        if (this._clearRafDefer) {
            this._clearRafDefer();
            this._clearRafDefer = null;
        }
    }
    clearState() {
        if (!this.app.isEnabled()) {
            nativeTimeout(() => {
                this.clearState();
            }, 600);
        }
        else {
            this.deactivate();
        }
    }
    deactivate() {
        this._clearDeferred();
        this._queue.length = 0;
        for (var i = 0; i < this._active.length; i++) {
            this._active[i].classList.remove(this._css);
        }
        this._active.length = 0;
    }
    _clearDeferred() {
        if (this._activeRafDefer) {
            this._activeRafDefer();
            this._activeRafDefer = null;
        }
    }
}
const ADD_ACTIVATED_DEFERS = 6;
const CLEAR_STATE_DEFERS = 6;
